File: concepts/004_resource_cleanup/nodejs/test_resource_cleanup.mjs
// 横向概念 004｜同步资源清理、异常安全与清理冲突。
// 共同问题：正常退出是否清理；异常退出是否清理；多个资源是否逆序清理；
// 清理由什么机制触发；清理失败与原始异常如何交互。
//
// polyglot-concept: resource_cleanup
// polyglot-related: languages/nodejs/language/test_027_explicit_resource_management_and_disposable_stacks.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

class RecordingResource {
  constructor(name, events, cleanupError = undefined) {
    this.name = name;
    this.events = events;
    this.cleanupError = cleanupError;
    events.push(`acquire:${name}`);
  }

  [Symbol.dispose]() {
    this.events.push(`release:${this.name}`);
    if (this.cleanupError !== undefined) {
      throw this.cleanupError;
    }
  }
}

test('using 在正常离开词法作用域时调用 Symbol.dispose', () => {
  const events = [];

  {
    using resource = new RecordingResource('normal', events);
    assert.equal(resource.name, 'normal');
    events.push('body');
  }

  assert.deepEqual(events, ['acquire:normal', 'body', 'release:normal']);
});

test('异常离开 using 作用域时仍先完成清理', () => {
  const events = [];
  const original = new Error('body failed');

  assert.throws(() => {
    using resource = new RecordingResource('exception', events);
    assert.equal(resource.name, 'exception');
    events.push('body');
    throw original;
  }, (error) => error === original);

  assert.deepEqual(events, ['acquire:exception', 'body', 'release:exception']);
});

test('DisposableStack 按注册的后进先出顺序清理资源', () => {
  const events = [];

  {
    using stack = new DisposableStack();
    stack.use(new RecordingResource('first', events));
    stack.use(new RecordingResource('second', events));
    events.push('body');
  }

  assert.deepEqual(events, [
    'acquire:first',
    'acquire:second',
    'body',
    'release:second',
    'release:first',
  ]);
});

test('Symbol.dispose 的返回值不能抑制代码块异常', () => {
  const original = new Error('body failed');

  assert.throws(() => {
    using resource = {
      [Symbol.dispose]() {
        return true;
      },
    };
    assert.ok(resource);
    throw original;
  }, (error) => error === original);

  // 与 Python __exit__ 不同，dispose 返回值被忽略；资源协议没有异常抑制通道。
});

test('清理错误与代码块错误组合为 SuppressedError', () => {
  const events = [];
  const bodyError = new Error('body failed');
  const cleanupError = new Error('cleanup failed');

  assert.throws(() => {
    using resource = new RecordingResource('failing', events, cleanupError);
    assert.equal(resource.name, 'failing');
    throw bodyError;
  }, (error) => {
    assert.equal(error instanceof SuppressedError, true);
    assert.equal(error.error, cleanupError);
    assert.equal(error.suppressed, bodyError);
    return true;
  });

  assert.deepEqual(events, ['acquire:failing', 'release:failing']);
});

File: concepts/03_errors_and_resources/01_resource_cleanup/python/test_resource_cleanup.py
"""横向概念 004｜同步资源清理、异常安全与清理冲突。

共同问题：正常退出是否清理；异常退出是否清理；多个资源是否逆序清理；
清理由什么机制触发；清理失败与原始异常如何交互。
"""

# polyglot-concept: resource_cleanup
# polyglot-related: languages/python/language/test_010_context_manager_protocols.py

import pytest


class RecordingResource:
    """只实现同步 ``with`` 协议，并记录取得与释放时点。"""

    def __init__(self, name, events, *, suppress=False, cleanup_error=None):
        self.name = name
        self.events = events
        self.suppress = suppress
        self.cleanup_error = cleanup_error
        self.exit_arguments = None

    def __enter__(self):
        self.events.append(f"acquire:{self.name}")
        return self

    def __exit__(self, exception_type, exception, traceback):
        self.events.append(f"release:{self.name}")
        self.exit_arguments = (exception_type, exception, traceback)
        if self.cleanup_error is not None:
            raise self.cleanup_error
        return self.suppress


def test_with_protocol_cleans_on_normal_exit():
    events = []
    resource = RecordingResource("normal", events)

    with resource as acquired:
        assert acquired is resource
        events.append("body")

    assert events == ["acquire:normal", "body", "release:normal"]
    assert resource.exit_arguments == (None, None, None)


def test_with_protocol_cleans_during_exception_exit():
    events = []
    resource = RecordingResource("exception", events)
    original = ValueError("body failed")

    with pytest.raises(ValueError) as caught:
        with resource:
            events.append("body")
            raise original

    assert caught.value is original
    assert events == ["acquire:exception", "body", "release:exception"]
    assert resource.exit_arguments[1] is original


def test_multiple_context_managers_clean_in_lifo_order():
    events = []

    with RecordingResource("first", events), RecordingResource("second", events):
        events.append("body")

    assert events == [
        "acquire:first",
        "acquire:second",
        "body",
        "release:second",
        "release:first",
    ]


def test_exit_true_can_suppress_the_body_exception():
    events = []

    with RecordingResource("suppressor", events, suppress=True):
        events.append("body")
        raise LookupError("handled")

    events.append("continued")
    assert events == ["acquire:suppressor", "body", "release:suppressor", "continued"]


def test_cleanup_error_replaces_body_error_and_preserves_context():
    events = []
    body_error = ValueError("body failed")
    cleanup_error = RuntimeError("cleanup failed")

    with pytest.raises(RuntimeError) as caught:
        with RecordingResource("failing", events, cleanup_error=cleanup_error):
            raise body_error

    assert caught.value is cleanup_error
    assert caught.value.__context__ is body_error
    assert events == ["acquire:failing", "release:failing"]
